lib/parsec/config.py

```diff
@@ -35,29 +35,32 @@ def __init__(self, msg):
         self.msg = 'ERROR: not a singular item: %s' % msg
 
 
-class config(object):
-    "Object wrapper for parsec functions"
-
-    def __init__(self, spec, upgrader=None, output_fname=None, tvars=None):
+class ParsecConfig(object):
+    """Object wrapper for parsec functions."""
 
+    def __init__(self, spec, upgrader=None, output_fname=None, tvars=None,
+                 validator=None):
         self.sparse = OrderedDictWithDefaults()
         self.dense = OrderedDictWithDefaults()
         self.upgrader = upgrader
         self.tvars = tvars
         self.output_fname = output_fname
-        self.checkspec(spec)
         self.spec = spec
-
-    def checkspec(self, spec, parents=None):
-        "check that the file spec is a nested dict of validators"
-        if not parents:
-            parents = []
-        for key, value in spec.items():
-            pars = parents + [key]
-            if isinstance(value, dict):
-                self.checkspec(value, pars)
-            else:
-                if not isinstance(value, validator):
+        if validator is None:
+            validator = parsec_validate
+        self.validator = validator
+
+    @staticmethod
+    def checkspec(spec_root, parents=None):
+        """Check that the file spec is a nested dict of specifications"""
+        stack = [[spec_root, []]]
+        while stack:
```

Comment:
To my mind, recursing through a nested dict is simpler as well as shorter than iteration, and (somewhat surprisingly!) it seems to be faster too. I think I asked you before (when your replaced another of my recursive implementations!) why you prefer iteration, but I can't recall the answer - can you remind me?

Reply:
Sorry it is really an educated preference. Recursion is easier to read but a bit more difficult to understand, at least for my brain!

There are several reasons for the preference. One is that iteration is easier to parallelize than recursion - should there be a necessity to do so. The other reason is my previous life being a Perl developer, where you can get warning for deep recursion even though the code is perfectly good. Deep recursion can be costly in memory as well - due to the function stack.

None applicable here, I guess. Happy to revert if you feel strongly about it.

Reply:
No, that's fine - I don't feel strongly enough about it to force a revert.